### PR TITLE
feat(persistence): span query DSL with SQL

### LINKED ISSUE #2081: Consistent partitioning and nesting by the `.` (dot) separator for span attributes
See [here](https://github.com/Arize-ai/phoenix/issues/2012#issuecomment-1888152790) and [here](https://github.com/Arize-ai/phoenix/issues/2012#issuecomment-1888159874) for more context.

Semantic conventions such as `llm.input_messages` [don't get partitioned](https://github.com/Arize-ai/phoenix/blob/3a7bfd2158b5c0ea995f4d1ec00a6f7b2c4a21e8/src/phoenix/trace/otel.py#L235) at the `.` (dot) separator, but non-semantic convention such as `llm.messages` do, and the latter creates a conflicting `llm` key in the attributes data structure on our internal spans.

The resulting attributes currently look like the one below. The inconsistent nesting means these two attribute values must be accessed via different patterns even though they share the same prefix.
```JavaScript
"attributes": {
    "llm": {
        "messages": ...,
    }, 
    "llm.input_messages": ...,
}
```
But ideally the attributes should look like the one below, where the nesting is consistent. (However, our filtering on semantic conventions currently doesn't work with this nested structure.)
```JavaScript
"attributes": {
    "llm": {
        "messages": ...,
        "input_messages": ...,
    },
}
```
This is problematic when users add arbitrary keys to the attributes that may share the same prefix as our semantic conventions. But because our semantic conventions get special treatment, this creates conflicting patterns for how attributes are accessed especially when two attributes may share the same prefixes.

The reason we have given special treatment to semantic convention is for backward compatibility. In particular, currently the way filtering is done assumes that the semantic conventions are not broken up. 

## Simultaneous downstream changes that are needed to make this work
- [ ] change how we filter on span attributes
- [ ] exporting to dataframe would need to flatten the nesting to reconstruct the current column structure

# Alternate Solution
## Eliminate the trie datastructure
We could also just not nest anything: the example above would become the following instead. We would still nest arrays when indices is encountered, e.g. `x.0.y`  becomes `{"x": [{"y": ...}]}`. But otherwise the attribute keys are unchanged. On the attribute value side, if we want anything to be a dictionary, which is a non-primitive type, then we would need to do `json.dumps` on the sender side and `json.loads` on the receiver side. This would require coordination, i.e. establishing some sort of convention for json strings.
```JavaScript
"attributes": {
    "llm.messages": ...,
    "llm.input_messages": ...,
}
```

### LINKED ISSUE #2780: [persistence] filters on spans

### LINKED ISSUE #2782: [persistence] explode and concat

### PR DESCRIPTION
**Pull request by RogerHYang**
resolves #2081
resolves #2780 
resolves #2782

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -143,6 +143,8 @@ dependencies = [
   "httpx", # For OpenAI testing
   "respx", # For OpenAI testing
   "nest-asyncio", # for executor testing
+  "astunparse; python_version<'3.9'",
```
Comment (axiomofjoy) [inline pyproject.toml:146]: A comment here would be helpful so we know why this one is necessary for 3.8

File: pyproject.toml [hunk 2]
```diff
@@ -143,6 +143,8 @@ dependencies = [
...
+  "sqlean.py",
```
Comment (axiomofjoy) [inline pyproject.toml:147]: Are we planning on adding `sqlean` as a stopgap until we figure out the conda situation?

File: src/phoenix/server/api/types/Project.py [hunk 3]
```diff
@@ -367,5 +369,5 @@ def validate_span_filter_condition(self, condition: str) -> ValidationResult:
             )
 
 
-LLM_TOKEN_COUNT_PROMPT = SpanAttributes.LLM_TOKEN_COUNT_PROMPT
-LLM_TOKEN_COUNT_COMPLETION = SpanAttributes.LLM_TOKEN_COUNT_COMPLETION
+LLM_TOKEN_COUNT_PROMPT = SpanAttributes.LLM_TOKEN_COUNT_PROMPT.split(".")
+LLM_TOKEN_COUNT_COMPLETION = SpanAttributes.LLM_TOKEN_COUNT_COMPLETION.split(".")
```
Comment (axiomofjoy) [inline src/phoenix/server/api/types/Project.py:373]: Doesn't this return a list? How does this work if we are using this as the key to a dictionary?

File: pyproject.toml (hunk 1 above)
Comment (RogerHYang) [reply pyproject.toml:146]: ok

File: pyproject.toml (hunk 2 above)
Comment (RogerHYang) [reply pyproject.toml:147]: yea, otherwise the unit tests won't pass on GitHub CI

File: src/phoenix/server/api/types/Project.py (hunk 3 above)
Comment (RogerHYang) [reply src/phoenix/server/api/types/Project.py:373]: yea, this is how JSON extract works in sqlchemy, i.e. by giving it a list

File: src/phoenix/server/api/types/Span.py [hunk 4]
```diff
@@ -306,39 +309,6 @@ def default(self, obj: Any) -> Any:
         return super().default(obj)
```
Comment (axiomofjoy) [inline src/phoenix/server/api/types/Span.py:309]: We don't need to address this here, but we should probably have this default to a safe encoding.

File: src/phoenix/utilities/__init__.py [hunk 5]
```diff
@@ -1,26 +0,0 @@
-from datetime import datetime
-from typing import List, Optional
-
-import pandas as pd
-
-from phoenix.core.project import Project
-from phoenix.trace.dsl import SpanQuery
-
-
-def query_spans(
```
Comment (axiomofjoy) [inline src/phoenix/utilities/__init__.py:10]: This get refactored?

File: src/phoenix/utilities/attributes.py
Comment (axiomofjoy) [inline src/phoenix/utilities/attributes.py:1]: Docstrings for the functions in this file would be helpful.

File: src/phoenix/utilities/__init__.py (hunk 5 above)
Comment (RogerHYang) [reply src/phoenix/utilities/__init__.py:10]: yes. this is moved from trace/otel.py and made into public functions

Comment (RogerHYang) [reply src/phoenix/utilities/__init__.py:10]: this is no longer needed

File: src/phoenix/session/client.py [hunk 6]
```diff
@@ -29,20 +28,15 @@ def __init__(
         self,
         *,
         endpoint: Optional[str] = None,
-        use_active_session_if_available: bool = True,
+        **kwargs: Any,  # for backward-compatibility
```
Comment (axiomofjoy) [inline src/phoenix/session/client.py:31]: We should add a deprecation warning if someone passes in `use_active_session_if_available`.

File: src/phoenix/session/session.py [hunk 7]
```diff
@@ -129,6 +127,77 @@ def __init__(
...
+        )
+
+    def query_spans(
+        self,
+        *queries: SpanQuery,
+        start_time: Optional[datetime] = None,
+        stop_time: Optional[datetime] = None,
+        root_spans_only: Optional[bool] = None,
+        project_name: Optional[str] = None,
+    ) -> Optional[Union[pd.DataFrame, List[pd.DataFrame]]]:
```
Comment (axiomofjoy) [inline src/phoenix/session/session.py:143]:
```suggestion
    def query_spans(
        self,
        *queries: SpanQuery,
        *,
        start_time: Optional[datetime] = None,
        stop_time: Optional[datetime] = None,
        root_spans_only: Optional[bool] = None,
        project_name: Optional[str] = None,
    ) -> Optional[Union[pd.DataFrame, List[pd.DataFrame]]]:
```

Comment (RogerHYang) [reply src/phoenix/session/session.py:143]: this is already implied by the `*` in `*queries:`

File: src/phoenix/db/models.py [hunk 8]
```diff
@@ -120,7 +125,15 @@ class Trace(Base):
     trace_id: Mapped[str]
     start_time: Mapped[datetime] = mapped_column(UtcTimeStamp, index=True)
     end_time: Mapped[datetime] = mapped_column(UtcTimeStamp)
-    latency_ms: Mapped[float]
+
+    @hybrid_property
+    def latency_ms(self) -> float:
+        return (self.end_time - self.start_time).total_seconds() * 1000
```
Comment (axiomofjoy) [inline src/phoenix/db/models.py:131]: Nice, I dig it.

Comment (RogerHYang) [reply src/phoenix/db/models.py:131]: the `.inplace.expression` part below this does that

Comment (RogerHYang) [reply src/phoenix/db/models.py:131]: basically, `classmethod` are db operations, while instance methods are python operations

File: src/phoenix/session/client.py (hunk 6 above)
Comment (RogerHYang) [reply src/phoenix/session/client.py:31]: ok

File: src/phoenix/utilities/attributes.py
Comment (RogerHYang) [reply src/phoenix/utilities/attributes.py:1]: ok

File: src/phoenix/utilities/attributes.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+    for item in sequence:
+        if isinstance(item, Mapping):
+            return True
+    return False
+
+
+def get_attribute_value(
+    attributes: Optional[Mapping[str, Any]],
+    key: str,
+    separator: str = ".",
```
Comment (axiomofjoy) [inline src/phoenix/utilities/attributes.py:99]: Do we need support for other separators?

Comment (RogerHYang) [reply src/phoenix/utilities/attributes.py:99]: no. this is just to make it explicit

File: src/phoenix/server/api/types/Span.py (hunk 4 above)
Comment (axiomofjoy) [reply src/phoenix/server/api/types/Span.py:309]: Good call.

File: src/phoenix/trace/dsl/query.py [hunk 10]
```diff
@@ -39,60 +52,29 @@
...
-            value = partial(self._from_attributes, key=key)
-        if self.key in _CONVERT_TO_STRING:
-            object.__setattr__(
-                self,
-                "value",
-                lambda span: None if (v := value(span)) is None else str(v),
-            )
-        else:
-            object.__setattr__(self, "value", value)
+        pass
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/query.py:65]: Is this needed if we just don't call `super().__post_init__()` in `_HasTmpSuffix`?

File: src/phoenix/trace/dsl/query.py [hunk 11]
```diff
@@ -105,61 +87,155 @@ def from_dict(cls, obj: Mapping[str, Any]) -> "Projection":
...
-                continue
-            record = (
-                {name: item.get(key) for name, key in self.kwargs.items()}
-                if self.kwargs
-                else dict(item)
+    def index_keys(self) -> List[str]:
+        return [self._primary_index.key, f"{self._position_prefix}position"]
+
+    def with_primary_index_key(self, _: str) -> "Explosion":
+        print("`.with_primary_index_key(...)` is deprecated and wil be removed in the future.")
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/query.py:140]:
```suggestion
        print("`.with_primary_index_key(...)` is deprecated and will be removed in the future.")
```

File: src/phoenix/trace/dsl/query.py [hunk 12]
```diff
@@ -105,61 +87,155 @@ def from_dict(cls, obj: Mapping[str, Any]) -> "Projection":
...
+        return self
+
+    def update_sql(self, sql: Select[Any], dialect: str) -> Select[Any]:
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/query.py:143]: Suggestion: You could use `Literal` here.

File: src/phoenix/trace/dsl/query.py [hunk 13]
```diff
@@ -105,61 +87,155 @@ def from_dict(cls, obj: Mapping[str, Any]) -> "Projection":
...
+
+    def _add_tmp_suffix(self, name: str) -> str:
+        if name.endswith(self._tmp_suffix):
+            return name
+        return name + self._tmp_suffix
+
+
+@dataclass(frozen=True)
+class Explosion(_HasTmpSuffix, Projection):
     kwargs: Mapping[str, str] = field(default_factory=lambda: MappingProxyType({}))
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/query.py:116]: What do `kwargs` represent here in terms of the query?

File: src/phoenix/trace/dsl/query.py [hunk 14]
```diff
@@ -105,61 +87,155 @@ def from_dict(cls, obj: Mapping[str, Any]) -> "Projection":
...
+                # Use zero-based indexing for backward-compatibility.
+                (position - 1).label(f"{self._position_prefix}position"),
+                *(
+                    (
+                        obj[key.split(".")].label(self._add_tmp_suffix(name))
+                        for name, key in self.kwargs.items()
+                    )
+                    if self.kwargs
+                    else (obj.label(self._array_tmp_col_label),)
+                ),
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/query.py:180]: This is pretty hard to follow. Maybe we can break it out into a separate line?

File: src/phoenix/trace/dsl/query.py [hunk 15]
```diff
@@ -268,75 +439,129 @@ def rename(self, **kwargs: str) -> "SpanQuery":
...
         return replace(self, _concat=_concat)
 
-    def with_explode_primary_index_key(self, primary_index_key: str) -> "SpanQuery":
-        _explode = self._explode.with_primary_index_key(primary_index_key)
-        return replace(self, _explode=_explode)
-
-    def __call__(self, spans: Iterable[Span]) -> pd.DataFrame:
+    def with_explode_primary_index_key(self, _: str) -> "SpanQuery":
+        print(
+            "`.with_explode_primary_index_key(...)` is deprecated and wil be "
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/query.py:450]:
```suggestion
            "`.with_explode_primary_index_key(...)` is deprecated and will be "
```

File: src/phoenix/trace/dsl/query.py [hunk 16]
```diff
@@ -404,3 +629,73 @@ def from_dict(
...
+            [
+                df.drop(attrs_label, axis=1),
+                df_attributes.add_prefix(attrs_label + "."),
+            ],
+            axis=1,
+        )
+    return df
+
+
+def _flatten_semantic_conventions(attributes: Mapping[str, Any]) -> Dict[str, Any]:
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/query.py:689]: Maybe an example would be helpful here.

File: tests/utilities/test_attributes.py [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+    actual = dict(unflatten(reversed(key_value_pairs)))
+    assert actual == desired
+
+
+@pytest.mark.parametrize("key_value_pairs,desired", [((("1.0.2", 0),), {"1": [{"2": 0}]})])
+def test_unflatten_separator(key_value_pairs, desired):
+    separator = str(random())
+    key_value_pairs = ((key.replace(".", separator), value) for key, value in key_value_pairs)
+    actual = dict(unflatten(key_value_pairs, separator=separator))
+    assert actual == desired
```
Comment (axiomofjoy) [inline tests/utilities/test_attributes.py:90]: For this test, can we avoid parameterization if we only have one pair? I do think it's preferable to avoid randomness in unit tests whenever possible (good discussion [here](https://softwareengineering.stackexchange.com/a/429603)). E.g., `separator = "-"` is better IMO.

File: tests/utilities/test_attributes.py [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+@pytest.mark.parametrize(
+    "key_value_pairs,desired",
+    [
+        ((), {}),
+        ((("1", 0),), {"1": 0}),
+        ((("1.2", 0),), {"1": {"2": 0}}),
+        ((("1.0.2", 0),), {"1": [{"2": 0}]}),
+        ((("1.0.2.3", 0),), {"1": [{"2": {"3": 0}}]}),
+        ((("1.0.2.0.3", 0),), {"1": [{"2": [{"3": 0}]}]}),
+        ((("1.0.2.0.3.4", 0),), {"1": [{"2": [{"3": {"4": 0}}]}]}),
```
Comment (axiomofjoy) [inline tests/utilities/test_attributes.py:16]: Is there some ambiguity here? How do I know that 4 is a key and not an index?

File: pyproject.toml (hunk 2 above)
Comment (axiomofjoy) [reply pyproject.toml:147]: Did you not add this to our main package dependencies because we might remove the dependency?

File: src/phoenix/trace/dsl/query.py [hunk 19]
```diff
@@ -404,3 +629,73 @@ def from_dict(
...
+    session: Session,
+    span_filter: SpanFilter,
+    project_name: str,
+    start_time: Optional[datetime] = None,
+    stop_time: Optional[datetime] = None,
+    root_spans_only: Optional[bool] = None,
+) -> pd.DataFrame:
+    # legacy labels for backward-compatibility
+    span_id_label = "context.span_id"
+    trace_id_label = "context.trace_id"
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/query.py:644]: Not part of this PR, but do we need to keep these around or are we ready to remove?

File: src/phoenix/trace/dsl/query.py [hunk 20]
```diff
@@ -404,3 +629,73 @@ def from_dict(
                 else {}
             ),
         )
+
+
+def _get_spans_dataframe(
+    session: Session,
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/query.py:635]: Suggestion: Pass in a `Connectable` rather than a `Session`.

File: tests/trace/dsl/test_query.py [hunk 21]
```diff
@@ -1,302 +1,708 @@
...
+        actual.sort_index().sort_index(axis=1),
+        expected.sort_index().sort_index(axis=1),
+    )
+    del sq, actual, expected
 
-SPAN_ID = "span_id"
-TRACE_ID = "trace_id"
 
+def test_select(session: Session) -> None:
+    sq = SpanQuery().select("name", tcp="llm.token_count.prompt")
```
Comment (axiomofjoy) [inline tests/trace/dsl/test_query.py:70]: On a first read here, I do not know what `tcp` stands for.

File: src/phoenix/trace/dsl/query.py [hunk 22]
```diff
@@ -268,75 +439,129 @@ def rename(self, **kwargs: str) -> "SpanQuery":
...
+        assert session.bind is not None
+        dialect = session.bind.dialect.name
+        conn = session.connection()
+        index = _NAMES[self._index.key].label(self._add_tmp_suffix(self._index.key))
+        row_id = models.Span.id.label(self._pk_tmp_col_label)
+        stmt = (
+            # We do not allow `group_by` anything other than `row_id` because otherwise
+            # it's too complex for the post hoc processing step in pandas.
+            select(row_id)
+            .join_from(models.Span, models.Trace)
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/query.py:483]: What is `join_from` doing that is different from `join`?

Comment (RogerHYang) [reply src/phoenix/trace/dsl/query.py:483]: oh this is actually redundant row and i forgot to change it.

File: tests/trace/dsl/test_query.py (hunk 21 above)
Comment (RogerHYang) [reply tests/trace/dsl/test_query.py:70]: yes...but that's also the point of the test, haha

File: src/phoenix/utilities/attributes.py
Comment (mikeldking) [reply src/phoenix/utilities/attributes.py:1]: Agree that this is probably the most complex part of this PR and probably requires the most "documentation"

File: src/phoenix/trace/dsl/query.py (hunk 13 above)
Comment (RogerHYang) [reply src/phoenix/trace/dsl/query.py:116]: In `.explode("retrieval.documents", content="document.content", score="document.score")` the `kwargs` are `content=` and `score=`.

File: tests/utilities/test_attributes.py [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+        ((("1.0.2.0.3.0.4", 0),), {"1": [{"2": [{"3": [{"4": 0}]}]}]}),
+        ((("1.2", 1), ("1", 0)), {"1": 0, "1.2": 1}),
+        ((("1.2.3", 1), ("1", 0)), {"1": 0, "1.2": {"3": 1}}),
+        ((("1.2.3", 1), ("1.2", 0)), {"1": {"2": 0, "2.3": 1}}),
+        ((("1.2.0.3", 1), ("1", 0)), {"1": 0, "1.2": [{"3": 1}]}),
+        ((("1.2.3.4", 1), ("1.2", 0)), {"1": {"2": 0, "2.3": {"4": 1}}}),
+        ((("1.0.2.3", 1), ("1.0.2", 0)), {"1": [{"2": 0, "2.3": 1}]}),
+        ((("1.2.0.3.4", 1), ("1", 0)), {"1": 0, "1.2": [{"3": {"4": 1}}]}),
+        ((("1.2.3.0.4", 1), ("1.2", 0)), {"1": {"2": 0, "2.3": [{"4": 1}]}}),
```
Comment (mikeldking) [inline tests/utilities/test_attributes.py:25]: this fixture is great but it might be worth having some realistic fixtures too - just so it's grocable in the context of attributes.

**Review by mikeldking — APPROVED** [review]

File: src/phoenix/trace/dsl/query.py (hunk 10 above)
Comment (RogerHYang) [reply src/phoenix/trace/dsl/query.py:65]: I tried it, and in that case the `__post_init__` on Projection doesn't run

File: src/phoenix/trace/dsl/query.py (hunk 13 above)
Comment (RogerHYang) [reply src/phoenix/trace/dsl/query.py:116]: it's the `kwargs` in `.explode(*args, **kwargs)`

File: tests/trace/dsl/test_query.py [hunk 24]
```diff
@@ -1,302 +1,768 @@
...
+    assert_frame_equal(
+        actual.sort_index().sort_index(axis=1),
+        expected.sort_index().sort_index(axis=1),
+    )
+    del sq, actual, expected
 
-SPAN_ID = "span_id"
-TRACE_ID = "trace_id"
 
+def test_select(session: Session) -> None:
```
Comment (axiomofjoy) [inline tests/trace/dsl/test_query.py:69]: Non-blocking suggestion: It would be nice to split these into separate these cases, since it looks like there are three distinct tests here. It's possible to group them under a class like `TestSelect` with method names like `test_select_token_count_prompt`. Easier to step through, and has the added benefit of providing clearer and more granular feedback when a test fails.

Comment (RogerHYang) [reply tests/trace/dsl/test_query.py:69]: good idea. i didn't realize you can do that

File: tests/utilities/test_attributes.py (hunk 18 above)
Comment (RogerHYang) [reply tests/utilities/test_attributes.py:16]: Yea, it is ambiguous. So in our methodology, 4 isn't considered an index here because we don't flatten lists unless each element is a dict, i.e. there are more keys to follow.

File: tests/utilities/test_attributes.py (hunk 17 above)
Comment (RogerHYang) [reply tests/utilities/test_attributes.py:90]: yea, that makes sense to me

File: src/phoenix/trace/dsl/query.py (hunk 12 above)
Comment (RogerHYang) [reply src/phoenix/trace/dsl/query.py:143]: good idea

File: src/phoenix/trace/dsl/query.py (hunk 14 above)
Comment (RogerHYang) [reply src/phoenix/trace/dsl/query.py:180]: ok

File: tests/trace/dsl/test_query.py (hunk 21 above)
Comment (axiomofjoy) [reply tests/trace/dsl/test_query.py:70]: Question for myself: Is there a reason "name" is passed as an argument and the other is passed as a kwarg?

File: src/phoenix/trace/dsl/query.py (hunk 19 above)
Comment (RogerHYang) [reply src/phoenix/trace/dsl/query.py:644]: I don't think we can remove these without breaking backward compatibility

File: src/phoenix/trace/dsl/query.py [hunk 25]
```diff
@@ -268,75 +439,129 @@ def rename(self, **kwargs: str) -> "SpanQuery":
...
+        if start_time:
+            stmt = stmt.where(start_time <= models.Span.start_time)
+        if stop_time:
+            stmt = stmt.where(models.Span.start_time < stop_time)
+        if root_spans_only:
+            parent = aliased(models.Span)
+            stmt = stmt.outerjoin(
+                parent,
+                models.Span.parent_id == parent.span_id,
+            ).where(parent.span_id == None)  # noqa E711
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/query.py:496]: Question for myself: How does this work?

File: tests/trace/dsl/test_query.py [hunk 26]
```diff
@@ -1,302 +1,768 @@
...
+            "name": ["root span"],
+            "Latency (milliseconds)": [30000.0],
+        }
+    ).set_index("context.span_id")
+    actual = sq(session, root_spans_only=True)
+    assert_frame_equal(
+        actual.sort_index().sort_index(axis=1),
+        expected.sort_index().sort_index(axis=1),
+    )
+    del sq, actual, expected
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply tests/trace/dsl/test_query.py:149]: It's just to clean up for the next test so it's not inadvertently re-using anything. Obviously it's not needed if there's only one test, but I left it there just in case (it's also just copied-pasted anyway).

File: tests/trace/dsl/test_query.py [hunk 27]
```diff
@@ -1,302 +1,768 @@
...
-            "document_position": 1,
-        },
+            "context.span_id": [],
+            "name": [],
+        }
+    ).set_index("context.span_id")
+    actual = sq(session, project_name="abc")
+    assert_frame_equal(
+        actual.sort_index().sort_index(axis=1),
+        expected.sort_index().sort_index(axis=1),
```
Comment (axiomofjoy) [inline tests/trace/dsl/test_query.py:237]: Probably don't need the sorts here.

File: tests/trace/dsl/test_query.py [hunk 28]
```diff
@@ -1,302 +1,768 @@
...
-            DOCUMENT_SCORE: 201,
-            "context.span_id": "2",
-            "document_position": 1,
-        },
+            "context.span_id": [],
+            "name": [],
+        }
+    ).set_index("context.span_id")
+    actual = sq(session, project_name="abc")
+    assert_frame_equal(
```
Comment (axiomofjoy) [inline tests/trace/dsl/test_query.py:235]: Might be simpler to assert the output dataframe is empty and has the expected columns.

File: tests/trace/dsl/test_query.py [hunk 29]
```diff
@@ -1,302 +1,768 @@
...
+        ),
+    )
+    assert_frame_equal(
+        actual.sort_index().sort_index(axis=1),
+        expected.sort_index().sort_index(axis=1),
+    )
+    del sq, actual, expected
+
+
+def test_filter_for_none(session: Session) -> None:
```
Comment (axiomofjoy) [inline tests/trace/dsl/test_query.py:220]: Same comment here about splitting into multiple test cases.

Comment (axiomofjoy) [reply tests/trace/dsl/test_query.py:220]: Looks like there are many tests on filtering in this file. Do these belong in `test_filter.py`?

Comment (RogerHYang) [reply tests/trace/dsl/test_query.py:220]: Yea they do, but they belong here too. So I opted to keep them here for more "holistic" testing and leave the more syntactical stuff over there.

File: src/phoenix/trace/dsl/filter.py [hunk 30]
```diff
@@ -1,199 +1,440 @@
...
-        translated = _Translator(condition).visit(root)
+        if not (source := self.condition):
+            return
+        root = ast.parse(source, mode="eval")
+        _validate_expression(root, source, valid_eval_names=self.valid_eval_names)
+        translated = _Translator(source).visit(root)
         ast.fix_missing_locations(translated)
         compiled = compile(translated, filename="", mode="eval")
         object.__setattr__(self, "translated", translated)
         object.__setattr__(self, "compiled", compiled)
```
Comment (axiomofjoy) [inline src/phoenix/trace/dsl/filter.py:88]: Curious, how robust is this to injections?

File: tests/utilities/test_attributes.py (hunk 23 above)
Comment (axiomofjoy) [reply tests/utilities/test_attributes.py:25]: Non-blocking suggestion: It would be helpful to use `pytest.param` with an ID to explicitly label a subset of these with a name that expresses the purpose of the test.

File: src/phoenix/trace/dsl/filter.py (hunk 30 above)
Comment (RogerHYang) [reply src/phoenix/trace/dsl/filter.py:88]: Should be robust because we are passing the result to sqlalchemy. If sqlalchemy is not robust then this is also not robust.

File: src/phoenix/utilities/attributes.py [hunk 31]
```diff
@@ -0,0 +1,278 @@
...
+)  # sorted so the longer strings go first
+
+
+def unflatten(
+    key_value_pairs: Iterable[Tuple[str, Any]],
+    *,
+    prefix_exclusions: Sequence[str] = (),
+    separator: str = ".",
+) -> Dict[str, Any]:
+    # `prefix_exclusions` is intended to contain the semantic conventions
```
Comment (axiomofjoy) [inline src/phoenix/utilities/attributes.py:56]: This comment belongs inside `_build_trie`.

File: src/phoenix/trace/dsl/query.py (hunk 10 above)
Comment (axiomofjoy) [reply src/phoenix/trace/dsl/query.py:65]: huh, weird

**Review by axiomofjoy — APPROVED** [review]
Really nice job, Roger. Stepping through the test cases helped me understand how the filters, query, and translator works.

File: tests/trace/dsl/test_query.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply tests/trace/dsl/test_query.py:149]: If it's needed, I think it's a sign that the test case should be split up. It's confusing to include it at the end of a test case when it's not needed, I think. If some cleanup process is needed, we should try to run it as a fixture when possible.
